libraries/AP_FETtecOneWire/AP_FETtecOneWire.cpp

```diff
@@ -0,0 +1,759 @@
+/*
+   This program is free software: you can redistribute it and/or modify
+   it under the terms of the GNU General Public License as published by
+   the Free Software Foundation, either version 3 of the License, or
+   (at your option) any later version.
+
+   This program is distributed in the hope that it will be useful,
+   but WITHOUT ANY WARRANTY; without even the implied warranty of
+   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+   GNU General Public License for more details.
+
+   You should have received a copy of the GNU General Public License
+   along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+
+/* Protocol implementation was provided by FETtec */
+
+#include <AP_Math/AP_Math.h>
+#include <AP_Scheduler/AP_Scheduler.h>
+#include <AP_SerialManager/AP_SerialManager.h>
+#include <SRV_Channel/SRV_Channel.h>
+#include <GCS_MAVLink/GCS.h>
+#include <AP_Arming/AP_Arming.h>
+
+#include "AP_FETtecOneWire.h"
+#if HAL_AP_FETTEC_ONEWIRE_ENABLED
+
+static constexpr uint32_t FTOW_UART_LOCK_KEY = 0x0FE77EC0;
+static constexpr uint32_t DELAY_TIME_US = 700;
+static constexpr uint32_t RESPONSE_LENGTH = 18;
+static constexpr uint8_t ALL_ID = 0x1F;
+static constexpr uint8_t MAX_TRANSMIT_LENGTH = 4;
+static constexpr uint8_t MAX_RECEIVE_LENGTH = 14;
+
+const AP_Param::GroupInfo AP_FETtecOneWire::var_info[] = {
+    // @Param: MASK
+    // @DisplayName: Channel Bitmask
+    // @Description: Enable of FETtec OneWire ESC protocol to specific channels
+    // @Bitmask: 0:Channel1,1:Channel2,2:Channel3,3:Channel4,4:Channel5,5:Channel6,6:Channel7,7:Channel8,8:Channel9,9:Channel10,10:Channel11,11:Channel12,12:Channel13,13:Channel14,14:Channel15,15:Channel16
+    // @RebootRequired: True
+    // @User: Standard
+    AP_GROUPINFO("MASK",  1, AP_FETtecOneWire, motor_mask, 0),
+
+    // @Param: POLES
+    // @DisplayName: Nr. electrical poles
+    // @Description: Number of motor electrical poles
+    // @Range: 2 50
+    // @RebootRequired: False
+    // @User: Standard
+    AP_GROUPINFO("POLES", 2, AP_FETtecOneWire, pole_count, 14),
+
+    AP_GROUPEND
+};
+
+AP_FETtecOneWire *AP_FETtecOneWire::_singleton;
+
+AP_FETtecOneWire::AP_FETtecOneWire()
+{
+    _singleton = this;
+
+    _response_length[OW_OK] = 1;

```

Comment:
These integer constants should be defines somewhere. I have no idea what they are for

Reply:
@teezeefpv I did not found a good way to get rid of those.

Reply:
The standard commands have a standard payload size. pull_command() uses it for transmitting and receiving the right amount of bytes depending on the command. Switch Case + defines could work, but would be worse that this.

Reply:
@andyp1per I did the change to a simple switch case statement, but it requires 104 bytes extra memory.
I do not think it is a good idea to change it.